internal/api/impl/v1/dashboard/validator/validator.go

```diff
@@ -0,0 +1,220 @@
+ */
+type Validator interface {
+	Process(dashboard *v1.Dashboard) error
+	LoadSchemas(path string)
+}
+
+type validator struct {
+	schemasConf config.Schemas
+	context     *cue.Context
+	baseDef     cue.Value
+	schemas     *sync.Map
+}
+
+const baseDefFileName = "base.cue"
+const pluginsFolderName = "plugins"
+
+/*
+ * Instanciate a validator
+ */
+func New(schemasConf config.Schemas) Validator {
+	return &validator{
+		schemasConf: schemasConf,
+		context:     cuecontext.New(),
+	}
+}
+
+/*
+ * Initialize the validator
+ * NB: this logic was part of New() before. It was extracted to avoid it being called by unrelated unit tests
+ */
+func (v *validator) initSchemas() {
+	// load the base panel definition
+	baseDefPath := fmt.Sprintf("%s/%s", v.schemasConf.Path, baseDefFileName)
+	data, err := os.ReadFile(baseDefPath)
+	if err != nil {
+		logrus.WithError(err).Errorf("Not able to read the base panel definition file at %s, shutting down..", baseDefPath)
+	}
+	baseDef := v.context.CompileBytes(data)
+
+	// initial load of plugins schemas
+	pluginsPath := fmt.Sprintf("%s/%s", v.schemasConf.Path, pluginsFolderName)
+	schemas, err := loadSchemas(v.context, baseDef, pluginsPath)
+	if err != nil {
+		logrus.WithError(err).Error("Not able to retrieve the list of schema files")
+	}
+
+	v.baseDef = baseDef
+	v.schemas = schemas
+}
+
+/*
+ * Validate a dashboard.
+ * The dashboard's panels are matched against the known list of CUE definitions (schemas).
+ * If no schema matches, the validation fails.
+ */
+func (v *validator) Process(dashboard *v1.Dashboard) error {
+	logrus.Tracef("Dashboard to validate: %+v", dashboard)
+
+	// init the validator if it wasn't already
+	if s, _ := v.baseDef.String(); len(s) == 0 {
+		v.initSchemas()
+	}
+
+	var res error
+
+	// go through the panels list
+	// the processing stops as soon as it detects an invalid panel TODO: can probably be improved
+	for k, panel := range dashboard.Spec.Panels {
+		logrus.Tracef("Panel to validate: %s", string(panel))
+
+		// compile the JSON panel into a CUE Value
+		value := v.context.CompileBytes(panel)
+
+		// retrieve panel's kind
+		kind, err := value.LookupPath(cue.ParsePath("kind")).String()
+		if err != nil {
+			logrus.Error("This panel doesn't contain the required \"kind\" field")
+			res = err
+			break
+		}
+
+		// retrieve the corresponding schema
+		schema, ok := v.schemas.Load(kind)
+		if !ok {
+			logrus.Errorf("%s is not valid panel: unknown kind %s", k, kind)
+			res = errors.New("Unknown panel kind")
+			break
+		}
+		logrus.Tracef("Matching panel %s against schema: %+v", k, schema)
+
+		// do the validation
+		unified := value.Unify(schema.(cue.Value))
+		opts := []cue.Option{
+			cue.Concrete(true),
+			cue.Attributes(true),
+			cue.Definitions(true),
+			cue.Hidden(true),
+		}
+		err = unified.Validate(opts...)
+		if err != nil {
+			logrus.Errorf("%s is not a valid %s panel", k, kind)
+			res = err
+			break
+		}
+	}
+
+	if res == nil {
```

Comment:
I'm not sure these logs are really interesting / useful.

At most, I would keep the debug log and remove the error one since the error has been log above.